File: algorithms/easy/count_elements_with_strictly_smaller_and_greater_elements.ex
defmodule Solution do
  @spec count_elements(nums :: [integer]) :: integer
  def count_elements(nums) do
    {smallest, num_of_smallest, biggest, num_of_biggest} = aux(nums)
    len = Enum.count(nums)

    if smallest == biggest do
      len - num_of_biggest
    else
      len - num_of_biggest - num_of_smallest
    end
  end

  def aux(nums) do
    aux(nums, 100_000, 0, -100_000, 0)
  end

  def aux([], smallest, num_of_smallest, biggest, num_of_biggest) do
    {smallest, num_of_smallest, biggest, num_of_biggest}
  end

  def aux([n | rest], smallest, num_of_smallest, biggest, num_of_biggest) do
    {smallest, num_of_smallest} =
      cond do
        smallest > n -> {n, 1}
        smallest == n -> {n, num_of_smallest + 1}
        true -> {smallest, num_of_smallest}
      end

    {biggest, num_of_biggest} =
      cond do
        biggest < n -> {n, 1}
        biggest == n -> {n, num_of_biggest + 1}
        true -> {biggest, num_of_biggest}
      end

    aux(rest, smallest, num_of_smallest, biggest, num_of_biggest)
  end
end

File: algorithms/easy/count_elements_with_strictly_smaller_and_greater_elements.rs
impl Solution {
    pub fn count_elements(nums: Vec<i32>) -> i32 {
        let mut smallest = i32::MAX;
        let mut num_of_smallest = 0;
        let mut biggest = i32::MIN;
        let mut num_of_biggest = 0;
        for &n in &nums {
            if smallest > n {
                smallest = n;
                num_of_smallest = 1;
            } else if smallest == n {
                num_of_smallest += 1;
            }

            if biggest < n {
                biggest = n;
                num_of_biggest = 1;
            } else if biggest == n {
                num_of_biggest += 1;
            }
        }

        if smallest == biggest {
            return nums.len() as i32 - num_of_biggest;
        }
        nums.len() as i32 - num_of_biggest - num_of_smallest
    }
}
